File: local_plugins/didit_sdk/ios/Classes/SdkFlutterPlugin.swift
import Flutter
import UIKit
import SwiftUI
@preconcurrency import DiditSDK

public class SdkFlutterPlugin: NSObject, FlutterPlugin {

    private var hostingController: UIViewController?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "didit_sdk", binaryMessenger: registrar.messenger())
        let instance = SdkFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startVerification":
            handleStartVerification(call, result: result)
        case "startVerificationWithWorkflow":
            handleStartVerificationWithWorkflow(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Start Verification with Token

    private func handleStartVerification(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let token = args["token"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Token is required", details: nil))
            return
        }

        let config = parseConfiguration(args["config"] as? [String: Any])

        DispatchQueue.main.async { [weak self] in
            self?.presentVerification(result: result) {
                DiditSdk.shared.startVerification(
                    token: token,
                    configuration: config
                )
            }
        }
    }

    // MARK: - Start Verification with Workflow

    private func handleStartVerificationWithWorkflow(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let workflowId = args["workflowId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Workflow ID is required", details: nil))
            return
        }

        let config = parseConfiguration(args["config"] as? [String: Any])

        DispatchQueue.main.async { [weak self] in
            self?.presentVerification(result: result) {
                DiditSdk.shared.startVerification(
                    workflowId: workflowId,
                    vendorData: args["vendorData"] as? String,
                    configuration: config
                )
            }
        }
    }

    // MARK: - Presentation Logic

    private func presentVerification(result: @escaping FlutterResult, startAction: @escaping () -> Void) {
        guard let rootVC = Self.findRootViewController() else {
            result([
                "type": "failed",
                "errorType": "unknown",
                "errorMessage": "Unable to find root view controller to present verification UI."
            ])
            return
        }

        let bridgeView = DiditBridgeView(
            onResult: { [weak self] verificationResult in
                let mapped = Self.mapVerificationResult(verificationResult)

                DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                    self?.hostingController?.dismiss(animated: false) {
                        self?.hostingController = nil
                    }
                }

                result(mapped)
            },
            startAction: startAction
        )

        let hostingVC = UIHostingController(rootView: bridgeView)
        hostingVC.modalPresentationStyle = .overFullScreen
        hostingVC.view.backgroundColor = .clear
        self.hostingController = hostingVC

        rootVC.present(hostingVC, animated: false)
    }

    // MARK: - View Controller Helpers

    private static func findRootViewController() -> UIViewController? {
        guard let windowScene = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .first,
              let rootVC = windowScene.windows.first(where: { $0.isKeyWindow })?.rootViewController
        else {
            return nil
        }

        var topVC = rootVC
        while let presented = topVC.presentedViewController {
            topVC = presented
        }
        return topVC
    }

    // MARK: - Configuration Parsing

    private func parseConfiguration(_ dict: [String: Any]?) -> DiditSdk.Configuration? {
        guard let dict = dict else { return nil }

        var language: SupportedLanguage?
        if let code = dict["languageCode"] as? String {
            language = SupportedLanguage.allCases.first { $0.code == code }
        }

        return DiditSdk.Configuration(
            languageLocale: language,
            fontFamily: dict["fontFamily"] as? String,
            loggingEnabled: dict["loggingEnabled"] as? Bool ?? false
        )
    }

    // MARK: - Result Mapping

    private static func statusString(_ status: VerificationStatus) -> String {
        switch status {
        case .approved: return "Approved"
        case .pending: return "Pending"
        case .declined: return "Declined"
        @unknown default: return "Pending"
        }
    }

    private static func mapVerificationResult(_ result: VerificationResult) -> [String: Any?] {
        switch result {
        case .completed(let session):
            return [
                "type": "completed",
                "sessionId": session.sessionId,
                "status": statusString(session.status)
            ]

        case .cancelled(let session):
            var dict: [String: Any?] = ["type": "cancelled"]
            if let session = session {
                dict["sessionId"] = session.sessionId
                dict["status"] = statusString(session.status)
            }
            return dict

        case .failed(let error, let session):
            var dict: [String: Any?] = [
                "type": "failed",
                "errorType": mapErrorType(error),
                "errorMessage": error.localizedDescription
            ]
            if let session = session {
                dict["sessionId"] = session.sessionId
                dict["status"] = statusString(session.status)
            }
            return dict

        @unknown default:
            return [
                "type": "failed",
                "errorType": "unknown",
                "errorMessage": "Unrecognized verification result"
            ]
        }
    }

    private static func mapErrorType(_ error: VerificationError) -> String {
        switch error {
        case .sessionExpired: return "sessionExpired"
        case .networkError: return "networkError"
        case .cameraAccessDenied: return "cameraAccessDenied"
        case .unknown: return "unknown"
        @unknown default: return "unknown"
        }
    }
}

// MARK: - SwiftUI Bridge View

private struct DiditBridgeView: View {
    let onResult: (VerificationResult) -> Void
    let startAction: () -> Void

    var body: some View {
        Color.clear
            .edgesIgnoringSafeArea(.all)
            .diditVerification { result in
                onResult(result)
            }
            .onAppear {
                startAction()
            }
    }
}

File: local_plugins/didit_sdk/android/src/main/kotlin/me/didit/sdk/sdk_flutter/SdkFlutterPlugin.kt
package me.didit.sdk.sdk_flutter

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import me.didit.sdk.Configuration
import me.didit.sdk.DiditSdk
import me.didit.sdk.DiditSdkState
import me.didit.sdk.SessionData
import me.didit.sdk.VerificationError
import me.didit.sdk.VerificationResult
import me.didit.sdk.VerificationStatus
import me.didit.sdk.core.localization.SupportedLanguage

class SdkFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var applicationContext: android.content.Context? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // ── FlutterPlugin ────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "didit_sdk")
        channel.setMethodCallHandler(this)
        applicationContext = binding.applicationContext

        if (!DiditSdk.isInitialized()) {
            DiditSdk.initialize(binding.applicationContext)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        applicationContext = null
    }

    // ── ActivityAware ────────────────────────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // ── MethodCallHandler ────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startVerification" -> handleStartVerification(call, result)
            "startVerificationWithWorkflow" -> handleStartVerificationWithWorkflow(call, result)
            else -> result.notImplemented()
        }
    }

    // ── Start Verification with Token ────────────────────────────────────────

    private fun handleStartVerification(call: MethodCall, result: Result) {
        val token = call.argument<String>("token")
        if (token == null) {
            result.error("INVALID_ARGUMENT", "Token is required", null)
            return
        }

        val config = parseConfiguration(call.argument<Map<String, Any>>("config"))

        scope.launch {
            try {
                DiditSdk.startVerification(
                    token = token,
                    configuration = config
                ) { verificationResult ->
                    result.success(mapVerificationResult(verificationResult))
                }

                awaitReadyAndLaunchUI(result)
            } catch (e: Exception) {
                resolveWithError(result, e)
            }
        }
    }

    // ── Start Verification with Workflow ──────────────────────────────────────

    private fun handleStartVerificationWithWorkflow(call: MethodCall, result: Result) {
        val workflowId = call.argument<String>("workflowId")
        if (workflowId == null) {
            result.error("INVALID_ARGUMENT", "Workflow ID is required", null)
            return
        }

        val config = parseConfiguration(call.argument<Map<String, Any>>("config"))

        scope.launch {
            try {
                DiditSdk.startVerification(
                    workflowId = workflowId,
                    vendorData = call.argument<String>("vendorData"),
                    configuration = config
                ) { verificationResult ->
                    result.success(mapVerificationResult(verificationResult))
                }

                awaitReadyAndLaunchUI(result)
            } catch (e: Exception) {
                resolveWithError(result, e)
            }
        }
    }

    // ── State Observation & UI Launching ──────────────────────────────────────

    private suspend fun awaitReadyAndLaunchUI(result: Result) {
        DiditSdk.state.first { state ->
            when (state) {
                is DiditSdkState.Ready -> {
                    val currentActivity = activity
                    if (currentActivity != null) {
                        DiditSdk.launchVerificationUI(currentActivity)
                    } else {
                        resolveWithError(
                            result,
                            IllegalStateException(
                                "No active Activity available to present verification UI."
                            )
                        )
                    }
                    true
                }
                is DiditSdkState.Error -> {
                    resolveWithError(
                        result,
                        IllegalStateException("Didit SDK reported an initialization error.")
                    )
                    true
                }
                else -> false
            }
        }
    }

    // ── Configuration Parsing ────────────────────────────────────────────────

    private fun parseConfiguration(map: Map<String, Any>?): Configuration? {
        if (map == null) return null

        var language: SupportedLanguage? = null
        val code = map["languageCode"] as? String
        if (code != null) {
            language = SupportedLanguage.fromCode(code)
        }

        return Configuration(
            languageLocale = language,
            fontFamily = map["fontFamily"] as? String,
            loggingEnabled = map["loggingEnabled"] as? Boolean ?: false
        )
    }

    // ── Result Mapping ───────────────────────────────────────────────────────

    private fun mapVerificationResult(result: VerificationResult): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()

        when (result) {
            is VerificationResult.Completed -> {
                map["type"] = "completed"
                putSessionData(map, result.session)
            }
            is VerificationResult.Cancelled -> {
                map["type"] = "cancelled"
                result.session?.let { putSessionData(map, it) }
            }
            is VerificationResult.Failed -> {
                map["type"] = "failed"
                map["errorType"] = mapErrorType(result.error)
                map["errorMessage"] = result.error.message ?: "An unknown error occurred."
                result.session?.let { putSessionData(map, it) }
            }
        }

        return map
    }

    private fun putSessionData(map: MutableMap<String, Any?>, session: SessionData) {
        map["sessionId"] = session.sessionId
        map["status"] = session.status.rawValue
    }

    private fun mapErrorType(error: VerificationError): String {
        return when (error) {
            is VerificationError.SessionExpired -> "sessionExpired"
            is VerificationError.NetworkError -> "networkError"
            is VerificationError.RetryBlocked -> "retryBlocked"
            is VerificationError.CameraAccessDenied -> "cameraAccessDenied"
            is VerificationError.NotInitialized -> "notInitialized"
            is VerificationError.ApiError -> "apiError"
            is VerificationError.Unknown -> "unknown"
        }
    }

    private fun resolveWithError(result: Result, e: Exception) {
        val errorResult = mapOf(
            "type" to "failed",
            "errorType" to "unknown",
            "errorMessage" to (e.message ?: "An unexpected error occurred.")
        )
        result.success(errorResult)
    }
}
